core/views.py

```diff
@@ -0,0 +1,256 @@
+from hacker import models as hacker_models
+
+
+class IndexView(base_views.TemplateView):
+
+    template_name = "index.html"
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        return context
+
+
+class SignupView(generic_views.FormView):
+
+    form_class = hacker_forms.SignupForm
+    template_name = "registration/signup.html"
+
+    def email_exists(self, form):
+        return hacker_models.Hacker.objects.filter(
+            email=form.cleaned_data.get("email")
+        ).exists()
+
+    def form_valid(self, form):
+        form.full_clean()
+        form.save()
+        return super(SignupView, self).form_valid(form)
+
+    def get(self, request, *args, **kwargs):
+        form = self.form_class(initial=self.initial)
+        return render(request, self.template_name, {"form": form})
+
+    # 'recipient_hacker' should be the `Hacker` model receiving the email
+    def send_confirmation_email(self, recipient_hacker):
+        recipient_hacker.generate_confirm_code()
+        to_email = getattr(recipient_hacker, "email", None)
+        confirm_code = getattr(recipient_hacker, "confirm_code", None)
+        first_name = getattr(recipient_hacker, "first_name", None)
+        if to_email is not None and confirm_code is not None:
+            email_content = (
+                "Almost there "
+                + first_name
+                + "! Use the following confirmation code to confirm your email: "
+                + confirm_code
+            )
+            send_mail(
+                "Confirm you email!",
+                email_content,
+                settings.EMAIL_HOST_USER,
+                [to_email],
+            )
+            hacker_models.Hacker.objects.filter(email=to_email).update(
+                confirm_code=confirm_code
+            )
+
+    def post(self, request, *args, **kwargs):
+        form = self.form_class(request.POST)
+        user_exists = False
+        if form.is_valid():
+            if not self.email_exists(form):
+                form.save()
+                username = form.cleaned_data.get("username")
+                raw_password = form.cleaned_data.get("password1")
+                user = authenticate(username=username, password=raw_password)
+                auth_login(request, user)
+                hacker = hacker_models.Hacker.objects.get(username=username)
+                # call the 'send_confirmation_email' function to send a confirmation email
+                self.send_confirmation_email(hacker)
+                return redirect(reverse_lazy("confirm_email"))
+            else:
+                user_exists = True
+        FormErrors = json.loads(form.errors.as_json())
+        return render(
+            request,
+            self.template_name,
+            {"form": form, "FormErrors": FormErrors, "user_exists": user_exists},
+        )
+
+
+# django.contrib.auth.LoginView
+class SignInView(generic_views.FormView):
```

Comment:
Don't use hand-rolled views for authentication. Django provides [built-in](https://docs.djangoproject.com/en/dev/topics/auth/default/#all-authentication-views) authentication views, and all you have to do is override the template if you want (we do want).

We don't even really need to override the class, we could include the authentication in `urls.py` with something like:

```
path('<your_url>', auth_views.LoginView.as_view(template_name='path/to/template')),
```
https://docs.djangoproject.com/en/dev/topics/auth/default/#all-authentication-views